lib/Dialect/TTIR/Analysis/DFShardingPolicy.cpp

```diff
@@ -0,0 +1,172 @@
+      // until they are exhausted from schedulable ops.
+      // TODO : We need to examine type of to_layout op and determine if for
+      // example we have a space in DRAM to perform this?(system->dram, double
+      // check this)
+      //
+      if (shardChainConfigs.back().isEmpty()) {
+        for (auto op : scheduleableOps) {
+          if (isa<ttir::ToLayoutOp>(op)) {
+            currentOp = op;
+            break;
+          }
+        }
+      }
+
+      if (currentOp == nullptr) {
+        currentOp = scheduleableOps[0];
+      }
+
+      // Schedule currentOp.
+      //
+      scheduler.scheduleOp(currentOp);
+
+      // Skip sharding process if currentOp is a ttir.to_layout op.
+      //
+      if (shardChainConfigs.back().isEmpty() &&
+          isa<ttir::ToLayoutOp>(currentOp)) {
+        currentOp = nullptr;
+        continue;
+      }
+
+      if (scheduler.hasUnscheduledOps()) {
+        scheduleableOps = scheduler.getScheduleableOps();
+
+        // Check if currentOp has a valid successor.
+        //
+        Operation *nextOp = nullptr;
+        for (auto op : scheduleableOps) {
+          for (auto operand : op->getOperands()) {
+            if (operand.getDefiningOp() == currentOp) {
+              nextOp = op;
+              break;
+            }
+          }
+        }
+
+        if (nextOp) {
+
+          // V0 Before adding to shard chain config check that currentOp is not
+          // fork/join op.
+          //
+          bool validForSharding = currentOp->hasOneUse() &&
+                                  legalGrids.lookup(currentOp).size() > 0 &&
+                                  legalGrids.lookup(nextOp).size() > 0;
+
+          // TODO
+          // It seems that bunch of TTNN ops have constraints which prevent
+          // them from being sharded if both inputs are interlieved, so my
+          // proposal is to try to resolve this by starting a shard chain with
+          // reshard op(at later phase only when necessary)
+          //
+
+          if (validForSharding) {
+            // Fetch largest legal sharded L1 grids for currentOp and nextOp.
+            //
+            // Calculate L1 tensor memory usage based on :
+            // currentOp output tensor shard spec, nextOp exec and nextOp output
+            // tensor.
+            //
+            LayoutAttr currentOpLayout = legalGrids.lookup(currentOp).front();
+            llvm::SmallVector<int64_t> currentOpShardShape =
+                currentOpLayout.getShardShape(false /*convertTileToScalar*/);
+            uint64_t l1InputUsage = currentOpLayout.getElementSizeBytes();
+            for (int64_t dim : currentOpShardShape) {
+              l1InputUsage *= dim;
+            }
+
+            LayoutAttr nextOpLayout = legalGrids.lookup(nextOp).front();
+            llvm::SmallVector<int64_t> nextOpShardShape =
+                nextOpLayout.getShardShape(false /*convertTileToScalar*/);
```

Comment:
we are assuming getShardShape would include (eventual) padding, right?

Reply:
I think it should include padding per layout spec.
BTW, I didn't plan to have this check permanently here, its a quick mock, it should be replaced with API call ~getL1ShardingMemUsage(input shards, output shard, op type).